File: Golang/Medium/1930_Unique_Length-3_Palindromic_Subsequences/solution.go
package uniquelength3palindromicsubsequences

func countPalindromicSubsequence(s string) int {
	letterSet := make(map[rune]bool)
	letterMap := make(map[rune][]int)

	for i, c := range s {
		if _, exist := letterSet[c]; !exist {
			letterSet[c] = true
			letterMap[c] = append(letterMap[c], i)
		} else {
			letterMap[c] = append(letterMap[c], i)
		}
	}

	count := 0
	for c, _ := range letterSet {
		startIndex := letterMap[c][0]
		endIndex := letterMap[c][len(letterMap[c])-1]

		betweenSet := make(map[rune]bool)
		for i := startIndex + 1; i < endIndex; i++ {
			betweenSet[rune(s[i])] = true
		}
		count += len(betweenSet)
	}
	return count
}

File: Python/Medium/1930_Unique_Length-3_Palindromic_Subsequences/solution.py
class Solution:
    def countPalindromicSubsequence(self, s: str) -> int:
        letterSet = set()
        letterMap = {}

        for i in range(len(s)):
            c = s[i]
            if c not in letterSet:
                letterSet.add(c)
                letterMap[c] = [i]
            else:
                letterMap[c].append(i)
        
        count = 0
        for c in letterSet:
            startIndex = letterMap[c][0]
            endIndex = letterMap[c][-1]

            betweenSet = set()
            for i in range(startIndex + 1, endIndex):
                betweenSet.add(s[i])
            
            count += len(betweenSet)
        return count
            
